File: phases/17-testing-and-verification/03-test-doubles-stubs-mocks-fakes-spies/code/main.ts
type User = {
  userId: string;
  preferredChannel: "email" | "sms";
};

class PreferenceStub {
  constructor(private readonly mapping: Record<string, "email" | "sms">) {}

  getChannel(userId: string): "email" | "sms" {
    return this.mapping[userId] ?? "email";
  }
}

class AuditFake {
  public rows: string[] = [];

  save(record: string): void {
    this.rows.push(record);
  }
}

class SenderSpy {
  public sentPayloads: string[] = [];

  send(channel: string, payload: string): void {
    this.sentPayloads.push(`${channel}:${payload}`);
  }
}

class StrictSenderMock {
  public calls = 0;

  constructor(private readonly expectedCorrelation: string) {}

  send(_channel: string, payload: string): void {
    if (!payload.includes(this.expectedCorrelation)) {
      throw new Error("missing required correlation ID");
    }
    this.calls += 1;
  }
}

class NotificationService {
  constructor(
    private readonly prefRepo: PreferenceStub,
    private readonly sender: { send: (channel: string, payload: string) => void },
    private readonly auditRepo: AuditFake,
  ) {}

  notify(user: User, message: string, correlationId: string): void {
    const channel = this.prefRepo.getChannel(user.userId);
    const payload = `cid=${correlationId};msg=${message}`;
    this.sender.send(channel, payload);
    this.auditRepo.save(`${user.userId}|${channel}|${payload}`);
  }
}

function assert(condition: boolean, message: string): void {
  if (!condition) throw new Error(message);
}

function runDemo(): void {
  const prefs = new PreferenceStub({ "u-1": "sms" });
  const spy = new SenderSpy();
  const audit = new AuditFake();
  const service = new NotificationService(prefs, spy, audit);

  service.notify({ userId: "u-1", preferredChannel: "sms" }, "deploy ok", "req-1");
  assert(spy.sentPayloads.length === 1, "expected one send");
  assert(audit.rows.length === 1, "expected one audit row");

  const strict = new StrictSenderMock("cid=req-2");
  const service2 = new NotificationService(new PreferenceStub({}), strict, new AuditFake());
  service2.notify({ userId: "u-2", preferredChannel: "email" }, "invoice", "req-2");
  assert(strict.calls === 1, "expected one strict mock call");
}

runDemo();
console.log("test doubles demo passed");

File: phases/17-testing-and-verification/03-test-doubles-stubs-mocks-fakes-spies/code/main.py
#!/usr/bin/env python3
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List


@dataclass(frozen=True)
class User:
    user_id: str
    preferred_channel: str


class PreferenceStub:
    def __init__(self, mapping: Dict[str, str]) -> None:
        self.mapping = mapping

    def get_channel(self, user_id: str) -> str:
        return self.mapping.get(user_id, "email")


class AuditFake:
    def __init__(self) -> None:
        self.rows: List[str] = []

    def save(self, record: str) -> None:
        self.rows.append(record)


class SenderSpy:
    def __init__(self) -> None:
        self.sent_payloads: List[str] = []

    def send(self, channel: str, payload: str) -> None:
        self.sent_payloads.append(f"{channel}:{payload}")


class StrictSenderMock:
    def __init__(self, expected_correlation: str) -> None:
        self.expected_correlation = expected_correlation
        self.calls = 0

    def send(self, channel: str, payload: str) -> None:
        if self.expected_correlation not in payload:
            raise AssertionError("missing required correlation ID")
        self.calls += 1


class NotificationService:
    def __init__(self, pref_repo: PreferenceStub, sender, audit_repo: AuditFake) -> None:
        self.pref_repo = pref_repo
        self.sender = sender
        self.audit_repo = audit_repo

    def notify(self, user: User, message: str, correlation_id: str) -> None:
        channel = self.pref_repo.get_channel(user.user_id)
        payload = f"cid={correlation_id};msg={message}"
        self.sender.send(channel, payload)
        self.audit_repo.save(f"{user.user_id}|{channel}|{payload}")


def demo_spy_and_fake() -> None:
    prefs = PreferenceStub({"u-1": "sms"})
    sender = SenderSpy()
    audit = AuditFake()
    svc = NotificationService(prefs, sender, audit)

    svc.notify(User("u-1", "sms"), "build green", "abc-123")

    assert len(sender.sent_payloads) == 1
    assert sender.sent_payloads[0].startswith("sms:")
    assert len(audit.rows) == 1


def demo_strict_mock_contract() -> None:
    prefs = PreferenceStub({"u-2": "email"})
    sender = StrictSenderMock(expected_correlation="cid=req-77")
    audit = AuditFake()
    svc = NotificationService(prefs, sender, audit)

    svc.notify(User("u-2", "email"), "invoice sent", "req-77")
    assert sender.calls == 1


def main() -> None:
    demo_spy_and_fake()
    demo_strict_mock_contract()
    print("test doubles demo passed")


if __name__ == "__main__":
    main()
